lineapy/graph_reader/artifact_collection.py

```diff
@@ -225,146 +206,57 @@ def _sort_session_artifacts(
-        }
-
     def _compose_module(
         self,
         session_artifacts_sorted: List[SessionArtifacts],
         indentation: int = 4,
+        return_dict_name="artifacts",
     ) -> str:
         """
         Generate a Python module that calculates artifacts
         in the given artifact collection.
         """
         indentation_block = " " * indentation
 
-        # Extract module script components by session
-        session_modules = [
-            self._extract_session_module(
-                session_artifacts=session_artifacts,
-                indentation=indentation,
-            )
-            for session_artifacts in session_artifacts_sorted
-        ]
-
-        # Combine components by "type"
         module_imports = "\n".join(
-            [module["session_imports"] for module in session_modules]
+            [
+                sa.get_session_module_imports()
+                for sa in session_artifacts_sorted
+            ]
         )
-        artifact_functions = "\n".join(
-            [module["artifact_functions"] for module in session_modules]
+
+        artifact_function_definition = "\n".join(
+            list(
+                itertools.chain.from_iterable(
+                    [
+                        sa.get_session_artifact_function_definitions(
+                            indentation=indentation
+                        )
+                        for sa in session_artifacts_sorted
+                    ]
+                )
+            )
         )
+
         session_functions = "\n".join(
-            [module["session_function"] for module in session_modules]
-        )
-        module_function_body = "\n".join(
-            [module["session_calculation"] for module in session_modules]
-        )
-        module_function_return = ", ".join(
-            [module["session_function_return"] for module in session_modules]
+            [
+                sa.get_session_function(indentation=indentation)
+                for sa in session_artifacts_sorted
+            ]
         )
 
-        input_parameters_body = []
-        for module in session_modules:
-            input_parameters_body += [
-                line.lstrip(" ").rstrip(",")
-                for line in module["session_input_parameters_body"].split("\n")
-                if len(line.lstrip(" ").rstrip(",")) > 0
+        module_function_body = "\n".join(
+            [
+                f"{indentation_block}{return_dict_name}.update({sa.get_session_function_callblock()})"
+                for sa in session_artifacts_sorted
             ]
+        )
 
-        module_input_parameters = [
-            InputVariable(
-                variable_name=x.split("=")[0].strip(" "),
-                value=x.split("=")[1].strip(" "),
+        module_input_parameters = []
+        for sa in session_artifacts_sorted:
+            module_input_parameters += list(
+                sa.get_session_input_parameters_spec().values()
             )
```

Comment:
nit: Maybe, use `extend()` instead, i.e., `module_input_parameters.extend(...)`?

Reply:
Is this faster?

Reply:
As per relevant online [discussion](https://stackoverflow.com/questions/3653298/concatenating-two-lists-difference-between-and-extend), it does not seem to make much difference in our case. I personally prefer `extend()` for readability purpose, but it's not a big deal.